### ISSUE TITLE
Bi-directional port mapping (Was: SNAT rule added (Possible dup #240))

### ISSUE THREAD

**Issue by neutronscott** [miniupnp/miniupnp#413]
Made new issue since #240 is old and I was not sure if same as the iptables rules were numerous.

I'd just like to see a SNAT rule added for each upnp port forward. For instance, multiple consoles (PS4, PC) playing Modern Warfare will ask for port udp/3074. It expects new outbound traffic from 3074 to be mapped to that new port (lets use 3148 in this example).

Currently new traffic inbound to 3148 does go to device:3074
but new traffic from device:3074 should be SNAT to 3148 so that NAT tests pass and multiple devices behind the same NAT router can host matches.

**Comment by neutronscott** [comment]
Here is how it looks manually inputted. I do not know of a way to SNAT without coding in WAN IP
```
-A PREROUTING -i eth0 -p udp -m udp --dport 3224 -j DNAT --to-destination 192.168.1.150:3074
-A PREROUTING -i eth0 -p udp -m udp --dport 3214 -j DNAT --to-destination 192.168.1.140:3074
-A POSTROUTING -s 192.168.1.150/32 -o eth0 -p udp -m udp --sport 3074 -j SNAT --to-source 76.x.x.x:3224
-A POSTROUTING -s 192.168.1.140/32 -o eth0 -p udp -m udp --sport 3074 -j SNAT --to-source 76.x.x.x:3214
```

**Comment by miniupnp** [comment]
how do we know the value to use for --to-source ?

**Comment by neutronscott** [comment]
would have to use ext_ip for this. I suppose configurations using ext_ifname could get IPs off the interfaces but certainly it wouldn't work with more than 1 available IP.
but nft may have another way. haven't learned it yet.

**Comment by neutronscott** [comment]
I have found that MASQUERADE before the catch-all one works good for this.

```
-I POSTROUTING -s 192.168.1.150/32 -o eth0 -p udp -m udp --sport 3074 -j MASQUERADE --to-ports 3224
```

This may help other applications but definitely is better for gamers.

**Comment by pali** [comment]
> Here is how it looks manually inputted. I do not know of a way to SNAT without coding in WAN IP
> 
> ```
> ...
> -A POSTROUTING -s 192.168.1.150/32 -o eth0 -p udp -m udp --sport 3074 -j SNAT --to-source 76.x.x.x:3224
> ...
> ```

One correcttion, there should be not external public IPv4 address, but IPv4 address of outgoing eth0 interface (in case of NAT 1:1 it can be also private IPv4 addres). And miniupnpd already knows this eth0 interface ip address.

Problem is that this address could change (e.g. after re-running dhcp client) and so, miniupnpd needs to handle eth0 inteface address changes and properly replace iptables rules.

Other option is to use `MASQUERADE` target instead of `SNAT` target. Documentation for `MASQUERADE` says:
> This target is only valid in the nat table, in the POSTROUTING chain. It should only be used with dynamically assigned IP (dialup) connections: if you have a static IP address, you should use the SNAT target. Masquerading is equivalent to specifying a mapping to the IP address of the interface the packet is going out, but also has the effect that connections are forgotten when the interface goes down. This is the correct behavior when the next dialup is unlikely to have the same interface address (and hence any established connections are lost anyway).
> 
> --to-ports port[-port]
This specifies a range of source ports to use, overriding the default SNAT source port-selection heuristics (see above). This is only valid if the rule also specifies one of the following protocols: tcp, udp, dccp or sctp.

So it is probably what miniupnpd can use as target itself finds correct interface IPv4 address.

> `-I POSTROUTING -s 192.168.1.150/32 -o eth0 -p udp -m udp --sport 3074 -j MASQUERADE --to-ports 3224`

**Comment by ptpt52** [comment]
in function `add_redirect_rule2()` https://github.com/miniupnp/miniupnp/blob/db82286683e81aa076eed5edaf543670b6ef5256/miniupnpd/netfilter/iptcrdr.c#L232

it assume that `iport == eport` always to be true on snat setup
that means no SNAT rule setup for MINIUPNPD on adding the redirect rule, this works if `iport == eport` always true on firewall SNAT.

yes it normally to be true, but it is not always the case.
for some SNAT firewall setup, like `-j MASQUERADE --random-fully`, that tranlate SNAT to random port, this case `iport != eport`, and so redirect rule add by `add_redirect_rule2()` here should be broken, and SNAT setup for MINIUPNPD is expected.

so I think we don't need to check `iport != eport`, but always setup SNAT for MINIUPNPD should do the trick.

NOTE: add_redirect_rule2() is called by `upnpc` by user, and user set iport equal to eport as args.

@miniupnp

**Comment by miniupnp** [comment]
@neutronscott what do you think about commit e49d44f700355552c45a95c1e067e4a815479557 ?

**Comment by Marc05** [comment]
> 
> 
> @neutronscott what do you think about commit [e49d44f](https://github.com/miniupnp/miniupnp/commit/e49d44f700355552c45a95c1e067e4a815479557) ?

I don't know what @neutronscott thinks, but I bet a **lot** of people would be happy to see this implemented, including myself!

**Comment by neutronscott** [comment]
I haven't tested it and I am not familiar with the code. The comments sound like it's going to do what I'd like.

**Comment by miniupnp** [comment]
@neutronscott @Marc05 is one of you able to test e49d44f ?

**Comment by Marc05** [comment]
I would love to! I'm not sure how I would go about that however. I'm happy to follow any instructions, or feel free to walk me through it (github@marc05.net).

From what I can tell, pfSense uses the latest release and applies patches to it to use in pfSense itself:
https://github.com/pfsense/FreeBSD-ports/tree/devel/net/miniupnpd

I also see that I can apply patches live on pfSense by using the Systems Patches package:
https://docs.netgate.com/pfsense/en/latest/development/system-patches.html

If we can make this fix as a patch that I can add through the System Patches package, I can test it on 2 different networks with different hardware.

**Comment by Firetop** [comment]
miniupnpd doesn't support its "masquerade" options on FreeBSD/pf like it does on Linux/netfilter. When it is set to masquerade, it sets up an outbound (snat/source nat) mapping to maintain a static source port for the same port requested by the client. Without that, you only get inbound NAT, and if you setup static port on outbound NAT you're only likely to get a single console working as expected.

**Comment by miniupnp** [comment]
I think to test you need to compile by yourself

**Comment by Firetop** [comment]
> I think to test you need to compile by yourself

I have sent an some messages around to the FreeBSD port maintainer of miniupnpd without any response yet, hopefully soon :) As soon as it's compiled I can test on my lab on physical hardware and on a VM as well.

https://github.com/freebsd/freebsd-ports/tree/master/net/miniupnpd

**Comment by Firetop** [comment]
> I think to test you need to compile by yourself

@miniupnp If I try to "gmake config.h" on freebsd after cloning, I get "makefile:38: ** missing separator. Stop"

**Comment by miniupnp** [comment]
@Firetop why are you using gmake ?

https://github.com/miniupnp/miniupnp/blob/master/miniupnpd/INSTALL#L11

**Comment by Firetop** [comment]
> @Firetop why are you using gmake ?
> 
> https://github.com/miniupnp/miniupnp/blob/master/miniupnpd/INSTALL#L11

1. I am using a FreeBSD VM for this, as I am not a professional at compiling code, I work in network support.

2. I cloned the git repo by using: git clone https://github.com/miniupnp/miniupnp.git /usr/src/

3. From there, I navigated to the following lactations where I figured a makefile would be: 

> A. cd /usr/src, 
> B. cd /usr/src/miniupnpd, 
> C. and cd /usr/src/miniupnpd/bsd. 

4. Tried both gmake and make with the same types of errors... Gmake was just cleaner to type out since I cannot copy and paste from the VM (Missing VMWare Tools).

5. I just retried with make config.h as the instructions mentioned (I did try that before) and it still gave me this error from the: /usr/src/miniupnpd location: 

```
make: "/usr/src/miniupnpd/makefile" line: 39: could not find bsdmake.inc
make: "/usr/src/miniupnpd/makefile" line 88: malformed conditional ($(FWNAME) == "pf")
make: "/usr/src/miniupnpd/makefile" line 160: warning "which bash || which ksh" returned non-zero status
make: fatal errors encountered -- cannot continue
```

6. If I try from the other directories, such as: /usr/src, or /usr/src/miniupnpd/bsd and received: 
`Don't know how to make config.h. Stop`

Any other ideas would be appreciated, so we can test this out and confirm this is what we need. I had also tried to use: genconfig.sh with no luck either. (Command not found)

FYI, I did find those instructions before you sent them, but clearly they could use a refresh with a bit of clarity on the setup required to compile for newbies, if multiple people are having issues compiling. (I know myself I sent this to 2 people I know and they got confused, and Marc05 doesn't know how to go about it as mentioned above.

**Comment by Firetop** [comment]
I think I am onto something..., Need to spinup a new VM, got make working on pfsense but no gcc...

**Comment by Firetop** [comment]
Okay! Was able to compile... It's installed on the VM though, anyway to move it to my pfsense install?

I was missing the following commands on FreeBSD 11.3.
(Which could maybe be added to your instructions)

Commands: 
pkg install gcc gmake wget
wget --no-check-certificate https://github.com/miniupnp/miniupnp/archive/master.zip
unzip master.zip
cd miniupnp-master/miniupnpd/
./genconfig.sh
make

Just wanted to confirm before I try applying this to prod. Currently the masquerade references are all inside their netfilter code which is only on Linux. Would this work for pfsense as well?

**Comment by Firetop** [comment]
I also recieved a response from the miniupnpd port mantainer for FreeBSD, he would like to know if  it would be enough to update to miniupnpd-2.1.20200329? Or is this an extra patch?

"I will look into this.
(Feel free to send me an e-mail again if you have not heard anything by
Thursday.)
Would it be enough to update to miniupnpd-2.1.20200329? Or is this an extra patch?"

Anymore feedback on the above would be helpful.

**Comment by miniupnp** [comment]
> I also recieved a response from the miniupnpd port mantainer for FreeBSD, he would like to know if it would be enough to update to miniupnpd-2.1.20200329? Or is this an extra patch?
> 
> "I will look into this.
> (Feel free to send me an e-mail again if you have not heard anything by
> Thursday.)
> Would it be enough to update to miniupnpd-2.1.20200329? Or is this an extra patch?"
> 
> Anymore feedback on the above would be helpful.

it is not yet in master branch.

**Comment by miniupnp** [comment]
@Firetop 
No need to put sources in /usr/src... Why not in your home directory ?

```
$ git clone  https://github.com/miniupnp/miniupnp.git
$ cd miniupnp/miniupnpd
$ ./genconfig.sh
$ make
$ make install
```

**Comment by miniupnp** [comment]
I'm stupid,  e49d44f is only for netfilter :( it will change nothing for pf :(

**Comment by miniupnp** [comment]
I don't know how I forgot this issue was about netfilter in the 1st place.

**Comment by Firetop** [comment]
> I don't know how I forgot this issue was about netfilter in the 1st place.

Trust me, we all have our days.. I didn't give much sleep last night so i apologize in advance if my train of thought was hard to follow, or a bit on the sharp side.

I have no issues simply to keep watch on this one.
And i'll re-try the compile once we know it will have an impact on PFsense?

**Comment by Marc05** [comment]
@miniupnp I'll forgive you if we can get masquerade on pf as well :D

People with pfSense have been dealing with this issue for years. All you have to do is google "pfsense upnp open nat"... and all of those issues come down to this! At least I've finally gotten some insight into the problem.

**Comment by Firetop** [comment]
> I don't know how I forgot this issue was about netfilter in the 1st place.

Did you need to submit a duplicate report for pfsense? Let me know. I know the title here mentions a possible duplicate already. So maybe one can be used for adding the masquerade function/features to pfsense since right now we are only modifying for Linux I believe :)

Let me know if any help on testing you may need. Again I'd be willing to submit a detailed git issue if a new one is infact needed for this change :)

**Comment by Marc05** [comment]
Any luck  @Firetop ?

@miniupnp  Should a separate request be submitted for the feature under pf and not just netfilter?

**Comment by miniupnp** [comment]
this issue is about netfilter. Please open another one if there is a problem with pf

**Comment by Firetop** [comment]
> Any luck @Firetop ?
> 
> @miniupnp Should a separate request be submitted for the feature under pf and not just netfilter?

Hey @Marc05 

@miniupnp would like us to submit a new issue, did you want to handle that incident report/request or should I? It will basically be a duplicate of this one, but for pfsense instead of netfilter. (Linux as I mentioned in my first post above).

More discussion on this with Jim over on the pfsense forum. We need this rolling, this should be classified as high priority when creating if you can.

If you rather I create it let me know. Otherwise drop the link to the new issue for us to check out after! :)

Thanks,
Joel S.

**Comment by Firetop** [comment]
> this issue is about netfilter. Please open another one if there is a problem with pf

Opened a new issue for pf specific.
Can be found here: https://github.com/miniupnp/miniupnp/issues/448

**Comment by miniupnp** [comment]
@neutronscott feel free to reopen if the issue is not fixed in the latest releases
